bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,149 @@
+* Flows for BNO to notify members about CorDapp version revocations. If a version of CorDapp was revoked, members are expected to manually update their nodes to the last not-revoked version ASAP.
+* Flows for BNO to collect reports from members about CorDapp versions installed on their nodes. Only versions of CorDapps related to *this* Business Network should be reported. CDS will rely on the information provided in CorDapp `MANIFEST` files until a better API is available.
+* Flows for members to manually request lists of available / revoked versions and their descriptions from BNO.
+* Flows for BNO to optionally notify members if they have an outdated or revoked version of CorDapp installed on their nodes.
+
+> It's important to emphasise, that in the Business Networks where members host nodes by themselves, BNO can only *do their best* to encourage members to upgrade by notifying them via CDS or sending them an email. Ultimately it will be up to a member to decide on whether they would like to upgrade or not. It should be in a member's best interest to promptly upgrade as otherwise they might loose ability to transact if a CorDapp is not backwards compatible.
+
+> CorDapps can be designed to stop working if a newer version is available. This can be done by making flows to compare the current version against the list of available versions from CDS. If the current version is not the latest - the flows might refuse to start. It will be up to the CorDapp developers to utilise such technics.
+
+#### High level architecture diagrams
+
+CDS can be used as a CorDapp or as a standalone library, with pluggable transports over HTTP or Corda flows. The following component architectures can be utilised:
+
+##### CDS as a CorDapp with a transport over HTTP
+
+![CDS as CorDapp](./resources/cds-as-cordapp-http.png)
+
+##### CDS as a CorDapp with a transport over Corda Flows
+![CDS as CorDapp](./resources/cds-as-cordapp-flows.png)
+
+##### CDS as a library over HTTP
+![CDS as library](./resources/cds-as-library-http.png)
+
+##### CDS as a library over Corda RPC
+![CDS as library](./resources/cds-as-library-rpc.png)
+
+There are no strict limitations around following one architecture or the other. BNs can mix and match depending on their requirements. While HTTP transport might seem to be an easier option, using Corda flows have a couple of benefits over it, in particular:
+* BNO will be able to host Maven in a private segment of their network, without exposing it to internet, effectively hidden behind firewalls and Corda.
+* BNO will be able to expose the repo to their BN members only by integrating `cds-cordapp` with Business Network Management Service.
+
+#### Implementation details
+
+The proposed solution will require implementation of the following components:
+
+* Custom Maven Resolver transports over flows and RPC. This will require the following interface to be implemented, which should be fairly straightforward to do:
```

Comment:
Yes sorta kinda. Problem is the need to annotate things with `@Suspendable` in Java 8. Once we upgrade to a newer Java and a newer quasar this goes away. Until then we might need to fork Maven Resolver and add some annotations ourselves - possibly also tweaking it to avoid checkpointing huge amounts of stuff (we can also potentially just not checkpoint at all and assume/ensure MR is idempotent)

Reply:
Great point. Let me reflect this.

Reply:
Is it possible to avoid checkpointing without making changes to Corda?